src/xenia/patcher/patch_db.cc

```diff
@@ -0,0 +1,164 @@
+        ReadPatchFile(patch_file.path / patch_file.name);
+    if (loaded_title_patches.title_id != -1) {
+      loaded_patches.push_back(loaded_title_patches);
+    }
+  }
+  XELOGI("PatchDB: Loaded patches for {} titles", loaded_patches.size());
+}
+
+PatchFileEntry PatchDB::ReadPatchFile(const std::filesystem::path& file_path) {
+  PatchFileEntry patchFile;
+  std::shared_ptr<cpptoml::table> patch_toml_fields;
+
+  try {
+    patch_toml_fields = cpptoml::parse_file(path_to_utf8(file_path));
+  } catch (...) {
+    XELOGE("PatchDB: Cannot load patch file: {}", path_to_utf8(file_path));
+    patchFile.title_id = -1;
+    return patchFile;
+  };
+
+  auto title_name = patch_toml_fields->get_as<std::string>("title_name");
+  auto title_id = patch_toml_fields->get_as<std::string>("title_id");
+  auto title_hash = patch_toml_fields->get_as<std::string>("hash");
+
+  patchFile.title_id = strtoul((*title_id).c_str(), NULL, 16);
+  patchFile.hash = strtoull((*title_hash).c_str(), NULL, 16);
+  patchFile.title_name = *title_name;
+
+  auto patch_table = patch_toml_fields->get_table_array("patch");
+
+  for (auto patch_table_entry : *patch_table) {
+    PatchInfoEntry patch = PatchInfoEntry();
+    auto patch_name = *patch_table_entry->get_as<std::string>("name");
+    auto patch_desc = *patch_table_entry->get_as<std::string>("desc");
+    auto patch_author = *patch_table_entry->get_as<std::string>("author");
+    auto is_enabled = *patch_table_entry->get_as<bool>("is_enabled");
+
+    patch.id = 0;  // Todo(Gliniak): Implement id for future GUI stuff
+    patch.patch_name = patch_name;
+    patch.patch_desc = patch_desc;
+    patch.patch_author = patch_author;
+    patch.is_enabled = is_enabled;
+
+    // Iterate through all available data sizes
+    for (const auto& type : patch_data_types_size) {
+      auto patch_entries = ReadPatchData(type.first, patch_table_entry);
+
+      for (const PatchDataEntry& patch_entry : *patch_entries) {
+        patch.patch_data.push_back(patch_entry);
+      }
+    }
+    patchFile.patch_info.push_back(patch);
+  }
+  return patchFile;
+}
+
+std::vector<PatchDataEntry>* PatchDB::ReadPatchData(
+    const std::string size_type,
+    const std::shared_ptr<cpptoml::table>& patch_table) {
+  std::vector<PatchDataEntry>* patch_data = new std::vector<PatchDataEntry>();
+  auto patch_data_tarr = patch_table->get_table_array(size_type);
+
+  if (!patch_data_tarr) {
+    return patch_data;
+  }
+
+  for (const auto& patch_data_table : *patch_data_tarr) {
+    uint32_t address = *patch_data_table->get_as<std::uint32_t>("address");
+    uint64_t value = *patch_data_table->get_as<std::uint64_t>("value");
+
+    if (size_type == "f32") {
+      float fvalue = float(*patch_data_table->get_as<double>("value"));
+      value = *(reinterpret_cast<uint32_t*>(&fvalue));
+    } else if (size_type == "f64") {
+      double dvalue = *patch_data_table->get_as<double>("value");
+      value = *(reinterpret_cast<uint64_t*>(&dvalue));
+    }
```

Comment:
Maybe this should be a switch case?

Reply:
Sure, but I would need to add some string->enum map while I load each patch